taichi/program/kernel.cpp

```diff
@@ -145,10 +142,97 @@ void Kernel::set_arg_int(int i, int64 d) {
   }
 }
 
+// NOTE: sync with snode.cpp: fetch_reader_result
+static uint64 fetch_result_uint64(int i)
+{
+  uint64 ret;
+  auto arch = get_current_program().config.arch;
+  if (arch == Arch::cuda) {
+    // TODO: refactor
+    // XXX: what about unified memory?
+#if defined(TI_WITH_CUDA)
+    CUDADriver::get_instance().memcpy_device_to_host(&ret,
+        (uint64 *)get_current_program().result_buffer + i,
+        sizeof(uint64));
+#else
+    TI_NOT_IMPLEMENTED;
+#endif
+  } else if (arch_is_cpu(arch)) {
+    ret = ((uint64 *)get_current_program().result_buffer)[i];
+  } else {
+    ret = get_current_program().context.get_arg_as_uint64(i);
+  }
+  return ret;
+}
+
+template <typename T>
+static T fetch_result(int i) // TODO: move to Program::fetch_result
```

Comment:
It would be great to move `fetch_result_uint64` and this to `Program` in this PR.  The modification would not be too big I guess, since you already need a few `get_current_program` calls in `fetch_result_uint64`.